From dc34c5e3989006a8a9f4de762ab0c5549ea09c7c Mon Sep 17 00:00:00 2001
From: Ivan Evlogiev <ivanevlogiev@codeaurora.org>
Date: Fri, 14 Feb 2014 16:05:22 +0200
Subject: [PATCH 109/176] QCamera2: Fix disabling flash during HDR

- Flash is disabled for HDR manual selection and for auto HDR positive
It may be enabled again only when HDR is disabled or for auto HDR negative

Change-Id: Iefb52c09b166b70b4ec019389b2b12d199502d29
CRs-Fixed: 612391
---
 QCamera2/HAL/QCameraParameters.cpp | 41 +++++++++++++++++++-------------------
 QCamera2/HAL/QCameraParameters.h   |  2 +-
 2 files changed, 22 insertions(+), 21 deletions(-)

diff --git a/QCamera2/HAL/QCameraParameters.cpp b/QCamera2/HAL/QCameraParameters.cpp
index 048e311..aacc3a3 100644
--- a/QCamera2/HAL/QCameraParameters.cpp
+++ b/QCamera2/HAL/QCameraParameters.cpp
@@ -2643,6 +2643,8 @@ int32_t QCameraParameters::setSceneMode(const QCameraParameters& params)
                 m_bHDREnabled = false;
             }
 
+            enableFlash(!m_bHDREnabled, false);
+
             if ((m_bHDREnabled) ||
                 ((prev_str != NULL) && (strcmp(prev_str, SCENE_MODE_HDR) == 0))) {
                 ALOGD("%s: scene mode changed between HDR and non-HDR, need restart", __func__);
@@ -2660,8 +2662,6 @@ int32_t QCameraParameters::setSceneMode(const QCameraParameters& params)
                     updateParamEntry(KEY_QC_HDR_NEED_1X, need_hdr_1x);
                 }
 
-                enableFlash(!m_bHDREnabled);
-
                 AddSetParmEntryToBatch(m_pParamBuf,
                                        CAM_INTF_PARM_HDR_NEED_1X,
                                        sizeof(m_bHDR1xFrameEnabled),
@@ -5449,13 +5449,7 @@ int32_t QCameraParameters::setHDRAEBracket(cam_exp_bracketing_t hdrBracket)
  *==========================================================================*/
 int32_t QCameraParameters::restoreAEBracket()
 {
-    int32_t rc = enableFlash(true);
-
-    if (NO_ERROR == rc) {
-      rc = setHDRAEBracket(m_AEBracketingClient);
-    }
-
-    return rc;
+  return setHDRAEBracket(m_AEBracketingClient);
 }
 
 /*===========================================================================
@@ -5465,22 +5459,26 @@ int32_t QCameraParameters::restoreAEBracket()
  *
  * PARAMETERS :
  *   @enableFlash : enable flash
+ *   @commitSettings : flag indicating whether settings need to be commited
  *
  * RETURN     : int32_t type of status
  *              NO_ERROR  -- success
  *              none-zero failure code
  *==========================================================================*/
-int32_t QCameraParameters::enableFlash(bool bflag)
+int32_t QCameraParameters::enableFlash(bool enableFlash, bool commitSettings)
 {
     int32_t rc = NO_ERROR;
-    if(initBatchUpdate(m_pParamBuf) < 0 ) {
-        ALOGE("%s:Failed to initialize group update table", __func__);
-        return BAD_TYPE;
+
+    if (commitSettings) {
+      if(initBatchUpdate(m_pParamBuf) < 0 ) {
+          ALOGE("%s:Failed to initialize group update table", __func__);
+          return BAD_TYPE;
+      }
     }
 
     const char *str = get(KEY_FLASH_MODE);
 
-    if (!bflag) {
+    if (!enableFlash) {
         str = FLASH_MODE_OFF;
     }
 
@@ -5496,7 +5494,8 @@ int32_t QCameraParameters::enableFlash(bool bflag)
                                       sizeof(value),
                                       &value);
         if (rc != NO_ERROR) {
-           ALOGE("%s:Failed to set led mode", __func__);
+          rc = BAD_VALUE;
+          ALOGE("%s:Failed to set led mode", __func__);
           return rc;
         }
     } else {
@@ -5504,10 +5503,12 @@ int32_t QCameraParameters::enableFlash(bool bflag)
         return rc;
     }
 
-    rc = commitSetBatch();
-    if (rc != NO_ERROR) {
-        ALOGE("%s:Failed to configure HDR bracketing", __func__);
-        return rc;
+    if (commitSettings) {
+      rc = commitSetBatch();
+      if (rc != NO_ERROR) {
+          ALOGE("%s:Failed to configure HDR bracketing", __func__);
+          return rc;
+      }
     }
 
     return rc;
@@ -6897,7 +6898,7 @@ void QCameraParameters::setHDRSceneEnable(bool bflag)
     m_HDRSceneEnabled = bflag;
 
     if (bupdate) {
-        enableFlash(!isHDREnabled());
+        enableFlash(!isHDREnabled(), true);
     }
 }
 
diff --git a/QCamera2/HAL/QCameraParameters.h b/QCamera2/HAL/QCameraParameters.h
index 39268f9..48717a6 100644
--- a/QCamera2/HAL/QCameraParameters.h
+++ b/QCamera2/HAL/QCameraParameters.h
@@ -499,7 +499,7 @@ public:
     bool isHDREnabled();
     bool isAutoHDREnabled();
     int32_t restoreAEBracket();
-    int32_t enableFlash(bool enableFlash);
+    int32_t enableFlash(bool enableFlash, bool commitSettings);
     int32_t updateRAW(cam_dimension_t max_dim);
     bool isAVTimerEnabled();
 
-- 
1.8.3.1

